Manifest V3 Support

johnsyweb opened the pull request:
#### Card

resolves #331 

#### Context

We don't want this extension to die with manifest.json v2.

#### Change

There's a lot.  See individual commits and comments for finer details.

#### Confirmation

##### Chrome

![Monosnap results | parkrun UK 2024-10-26 13-32-22](https://github.com/user-attachments/assets/30cdb456-9a02-457d-8f19-abcd3fc551b7)

##### Firefox

<img width="1440" alt="Monosnap results | parkrun UK 2024-10-26 13-30-11" src="https://github.com/user-attachments/assets/44b89b22-08de-4032-a471-22c0a1c8218e">

johnsyweb commented on build/extension-chrome/build.sh:
```diff
@@ -46,9 +46,6 @@ cp -r browser-extensions/common/css ${TMP_BUILD_DIR}/
 
-# Copy the extras libraries and code for Chrome
-cp -r browser-extensions/chrome/js ${TMP_BUILD_DIR}/
-
```
The polyfill has gone.

johnsyweb commented on build/version.sh:
```diff
@@ -1,6 +1,6 @@
 #!/bin/bash -xe
 
-export VERSION=1.4.0
+export VERSION=2.0.0
```
I figure this is a big enough change to warrant a major version.

johnsyweb commented on browser-extensions/common/js/lib/challenges_ui.js:
```diff
@@ -370,7 +360,7 @@ function createVoronoiMapPrototype() {
       // https://github.com/zetter/voronoi-maps/blob/master/lib/voronoi_map.js
 
-      var cell_group = document.createElement("g")
+      var cell_group = document.createElementNS('http://www.w3.org/2000/svg', "g")
```
https://webhint.io/docs/user-guide/hints/hint-create-element-svg/

johnsyweb reviewed (commented):
🙏

johnsyweb commented on browser-extensions/firefox/manifest.json:
```diff
@@ -70,83 +69,110 @@
-        "js/lib/i18n.js",
-        "js/background.js"
-    ]
+    "service_worker": "js/background.js",
```
☝️ This is ignored by Firefox but needed by Chrome.
👇 This is rejected by Chrome but needed by Firefox. 🤦

fraz3alpha replied:
Currently for firefox we have:
```
  "background": {
    "service_worker": "js/background.js",
    "scripts": ["js/background.js"]
  }
```
can we just remove `service_worker` in the Firefox specific manifest file? 

I know it would be different to the Chrome one, but if it works, that's fine.

oliverdunk replied:
Starting in Chrome 121, you should be fine to specify both: https://developer.chrome.com/docs/extensions/whats-new#background-key-warning. Chrome will emit a warning for the `scripts` key but it doesn't block installation and won't show to users when published to the Chrome Web Store.

fraz3alpha commented:
Thank you so much for doing this @johnsyweb - I'll try and find time as soon as I can to review it properly!

oliverdunk commented:
Hey, I work on the Chrome Extensions team and just found this after installing the extension (I did my first parkrun today!). It looks like you have everything under control but feel free to reach out to me if you need any help with the migration. Keep up the great work! 🙌

fraz3alpha commented on browser-extensions/chrome/manifest.json:
```diff
@@ -1,61 +1,61 @@
       ],
       "js": [
-          "js/lib/version.js",
-          "js/lib/third-party/browser-polyfill/browser-polyfill-76eeeac.js",
```
The browser polyfill allowed us to use the same `promises` in Firefox and Chrome - as I recall it wasn't to do with trying to find out which browser we were on. Are they both equivalent now without this?

fraz3alpha replied:
It seems to work, so I guess it's fine :)

fraz3alpha commented:
# Testing

## Chrome

### Build

It build via the Docker build

### Loading + user settings

Loaded OK with no errors.

No problems with initially setting the user settings (ID + home parkrun)

### Console logs loading https://www.parkrun.org.uk/parkrunner/1309364/all/

Shows this in the console:
```
Denying load of chrome-extension://fipfmmhmenaheehlodlmfcapmhbdmgeh/css/third-party/leaflet-fullscreen/images/fullscreen@2x.png. Resources must be listed in the web_accessible_resources manifest key in order to be loaded by pages outside the extension.
```
```
all/:1      
GET chrome-extension://invalid/ net::ERR_FAILED
```

## Firefox

### Build

It built via the Docker build.

### Loading + user settings

Loading it through the temporary add-on route (because it isn't signed) - https://extensionworkshop.com/documentation/develop/temporary-installation-in-firefox/, it gave these errors:

![image](https://github.com/user-attachments/assets/d8fb3486-3b8e-45ec-b9e1-7428479456ff)

No problems with initially setting the user settings (ID + home parkrun)

### Console logs loading https://www.parkrun.org.uk/parkrunner/1309364/all/

![image](https://github.com/user-attachments/assets/0bc26745-684f-4f42-9321-2453bd3b04ee)

These must be parkrun cookies, as they are nothing to do with us.

There are far fewer logs in Firefox though... which is curious, the screenshot is filtered to errors, but even selecting debug there is very little, and it's not Chrome.

johnsyweb commented:
Hmmm… will take a closer look at 1313de282876802585367d53b105ad933a115594 when I get some time.

fraz3alpha commented:
# Investigating the images that fail to load

Here is the block that allows resources to be loaded:
```
      "resources": [
        "/images/badges/*.png",
        "/images/logo/*.png",
        "/images/maps/markers/*.png",
        "/images/flags/*.png",
        "/html/*.html",
        "/css/*"
      ],
```

I have tried adding these lines, just in case it doesn't like nested directories:
```
        "/css/third-party/leaflet-extramarkers/images/*.png",
        "/css/third-party/leaflet/images/*.png",
        "/css/third-party/leaflet-fullscreen/images/*.png"
```
but that didn't work.

Then I checked that removing ` "/images/badges/*.png",` caused lots of things to break - and it did - so it's definitely the right block of code, and some lines are definitely doing something. Also, that shows I can edit the manifest in the unpacked extension directory, and just click reload and that will refresh this list.

I tried putting my new, more specific, lines at the top, in case it only allows 6 entries or something random, but that made no difference
```
      "resources": [
        "/css/third-party/leaflet-extramarkers/images/*.png",
        "/css/third-party/leaflet/images/*.png",
        "/css/third-party/leaflet-fullscreen/images/*.png",
        "/images/badges/*.png",
        "/images/logo/*.png",
        "/images/maps/markers/*.png",
        "/images/flags/*.png",
        "/html/*.html",
        "/css/*"
      ],
```

It still says
```
all/:1 Denying load of chrome-extension://fipfmmhmenaheehlodlmfcapmhbdmgeh/css/third-party/leaflet-fullscreen/images/fullscreen@2x.png.
```

This comes in via this CSS:
```
@media
  (-webkit-min-device-pixel-ratio:2),
  (min-resolution:192dpi) {
    .leaflet-control-fullscreen a {
      background-image: url('chrome-extension://__MSG_@@extension_id__/css/third-party/leaflet-fullscreen/images/fullscreen@2x.png');
    }
  }
```

note, that this is one of the files we patch, so that it loads the image from the extension path, and not a path on the server. The patch looks like this:
```
@@ -35,6 +35,6 @@
   (-webkit-min-device-pixel-ratio:2),
   (min-resolution:192dpi) {
     .leaflet-control-fullscreen a {
-      background-image: url('images/fullscreen@2x.png');
+      background-image: url('chrome-extension://__MSG_@@extension_id__/css/third-party/leaflet-fullscreen/images/fullscreen@2x.png');
     }
   }
```

I can always load `chrome-extension://fipfmmhmenaheehlodlmfcapmhbdmgeh/css/third-party/leaflet-fullscreen/images/fullscreen@2x.png` in the browser whether or not the `/css` path in tin the resources section of the manifest - so perhaps it doesn't go through the same checking when it isn't embedded in a webpage.

oliverdunk commented:
> all/:1 Denying load of chrome-extension://fipfmmhmenaheehlodlmfcapmhbdmgeh/css/third-party/leaflet-fullscreen/images/fullscreen@2x.png.

Can you try removing `use_dynamic_url` from the `web_accessible_resources` key for those resources?

When that is set to true, the CSS can only be loaded by a website by using a special dynamic URL generated by `chrome.runtime.getURL`. For example, `chrome-extension://ABC-DEF-GHI/css/third-party/leaflet-fullscreen/images/fullscreen@2x.png`.

This is a different ID to the one that you normally use and different to the one that `__MSG_@@extension_id__` in your CSS file is replaced with.

If that works, it is probably fine in this case to ship as is. The purpose of the dynamic URLs is to prevent fingerprinting where a website attempts to request that resource and uses the status code to determine if your extension is installed. If you'd like to preserve that, you might want to look at rendering the map in an iframe rather than injecting the JS and CSS directly with a content script. Inside of an iframe, you can load as normal with the static extension ID.

fraz3alpha commented on browser-extensions/chrome/manifest.json:
```diff
@@ -1,61 +1,61 @@
+        "js/lib/third-party/leaflet-canvasicon/leaflet-canvasicon-0.1.4.js",
+        "js/lib/third-party/leaflet-piechart/leaflet-piechart-0.1.2.js",
+        "js/lib/third-party/d3-voronoi/d3-voronoi.js",
+        "js/lib/cache.js",
```
Seems this file has been added twice to this list, I'll remove one

johnsyweb commented:
I wonder whether I broke the calls to `runtime.getURL()` when I removed the polyfill and replaced it with a more rudimentary method to get the `browser` / `chrome` instance.

If we're removing the dynamic URLs (which don't seem entirely necessary), we can probably remove the getURL calls, making the code simpler and more consistent across browsers.

I hope to get some time after parkrun today  to investigate.

fraz3alpha commented:
@johnsyweb  - it's actually fine, it works now. You still need to get the path to the resources in the extension, and that `getURL()` is working - in fact it was the thing that was loading the images right, it was only the CSS that couldn't load the marker and icon images for the maps - all the badges etc.. were fine.

So, nothing really to look into, I've got it working - the only thing that was missing was the bits relating to the maps which were via CSS, and that seems good enough - it sounds like @oliverdunk was correct in that even though both mechanisms generated the same URL, the javascript one must have been magic and included some back end shenanigans to make it available, whereas the CSS one didn't do that, so it didn't work.

fraz3alpha commented on browser-extensions/common/js/lib/challenges.js:
```diff
@@ -2176,11 +2122,9 @@ function challenge_tourist(data, params) {
 function is_our_page(data) {
-  // return has_user_data_athlete_id(data) && has_this_athlete_id(data) && get_user_data_athlete_id(data) == get_this_athlete_id(data) 
-  return (data.info !== undefined && data.info.is_our_page)
+  return data?.info?.is_our_page ?? false
```
Getting fancy with Javascript now :)

johnsyweb replied:
Ha. It was this extension that encouraged me to learn the language (https://exercism.org/tracks/javascript), so thank you!

fraz3alpha reviewed (approved):
I've finally got through all of this, and it looks good!

I've got a few tweaks in another PR, but I propose I merge this one, and then merge mine into the main branch as well - and we'll get it sorted out.

I can't thank you enough for you help here @johnsyweb - I haven't had enough free time to look at this, so I'm really grateful you were able to get it working!

johnsyweb commented:
Wonderful. I'm glad to see this was helpful. It was a fun exercise.
